### PR TITLE
[builtin] Add a new SIL builtin convertStrongToUnsafeUnowned()

### PR DESCRIPTION
**Pull request by gottesmm**
The signature is:

(T, @inout @unowned(unsafe) T) -> ()

The reason for the weird signature is that currently the Builtin infrastructure
does not handle guaranteed results well.

The semantics of this builtin is that it enables one to store the first argument
into an unowned unsafe address without any reference counting operations. It
does this just by SILGening the relevant code. The optimizer chews through this
code well, so we get the expected behavior.

I also included a small proof of concept to validate that this builtin works as
expected.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: lib/SIL/OperandOwnership.cpp [hunk 1]
```diff
@@ -1033,6 +1033,12 @@ CONSTANT_OWNERSHIP_BUILTIN(Owned, MustBeInvalidated, UnsafeGuaranteed)
 #define BUILTIN(X, Y, Z)
 #include "swift/AST/Builtins.def"
 
+OperandOwnershipKindMap
+OperandOwnershipKindBuiltinClassifier::visitConvertStrongToUnownedUnsafe(
+    BuiltinInst *, StringRef) {
+  llvm_unreachable("Builtin should have been lowered to SIL instruction?!");
+}
```
Comment (jckarter) [inline lib/SIL/OperandOwnership.cpp]: Wouldn't the BUILTIN_SIL_OPERATION expansion handle this already?

File: lib/SILGen/LValue.h [hunk 2]
```diff
@@ -405,7 +405,11 @@ class LValue {
     assert(&component == Path.back().get());
     Path.pop_back();
   }
-  
+
+  /// Pop the last component off this LValue unsafely. Please be careful when
+  /// using this!
+  void unsafelyDropLastComponent() & { Path.pop_back(); }
```
Comment (jckarter) [inline lib/SILGen/LValue.h]: Can we check whether that last component is the expected ownership component when we drop it?

File: lib/SILGen/SILGenBuiltin.cpp [hunk 3]
```diff
@@ -1045,6 +1045,70 @@ emitBuiltinGlobalStringTablePointer(SILGenFunction &SGF, SILLocation loc,
...
+        << "lhsType: " << objectSrcValue->getType() << "\n"
+        << "rhsType: " << inoutDest->getType() << "\n";
+    llvm::report_fatal_error("standard fatal error msg");
+  }
+
+  // Ok. We have the right types. First convert objectSrcValue to its
+  // unowned representation.
+  SILType optionalType = SILType::getOptionalType(objectSrcValue->getType());
+  SILValue someVal =
+      SGF.B.createOptionalSome(loc, objectSrcValue, optionalType);
```
Comment (jckarter) [inline lib/SILGen/SILGenBuiltin.cpp]: Should we also handle non-optional unowned values?

Comment (gottesmm) [reply lib/SILGen/SILGenBuiltin.cpp]: That might make sense. But I want to do that later if so.

File: lib/SIL/OperandOwnership.cpp (hunk 1 above)
Comment (gottesmm) [reply lib/SIL/OperandOwnership.cpp]: I was following the example of GlobalStringTablePointer. Let me see.

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Review by jckarter — APPROVED** [review]

**Pull request merged**